controllers/bookController.go

```diff
@@ -0,0 +1,219 @@
+	_, err := es.Client.Delete().
+		Index(es.Index).
+		Type(es.DocType).
+		Id(id).
+		Do(ctx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e ElasticBookController) Search(fields map[string]string) ([]models.Book, error) {
+
+	query := elastic.NewBoolQuery()
+
+	if title, ok := fields["title"]; ok && title != "" {
+		titleMatchQuery := elastic.NewMatchQuery("title", title)
+
+		if author, ok := fields["author_name"]; ok && author != "" {
+			authorMatchQuery := elastic.NewMatchQuery("author_name", author)
+			query.Must(titleMatchQuery, authorMatchQuery)
+
+		} else {
+			query.Must(titleMatchQuery)
+		}
+
+	} else if author, ok := fields["author_name"]; ok && author != "" {
+		authorMatchQuery := elastic.NewMatchQuery("author_name", author)
+		query.Must(authorMatchQuery)
+	}
+	if price, ok := fields["price_range"]; ok && price != "" {
+		var from, to int
+		var err error
+		s := strings.Split(price, "-")
+		if len(s) != 2 {
+			return nil, models.NewStringError("Conversion failed for `Price_range` field")
+		}
+		if from, err = strconv.Atoi(s[0]); err != nil {
+			return nil, models.NewStringError("Conversion failed for `Price_range` field")
+		}
+		if to, err = strconv.Atoi(s[1]); err != nil {
+			return nil, models.NewStringError("Conversion failed for `Price_range` field")
+		}
+		priceFilterQuery := elastic.NewRangeQuery("price").From(from).To(to)
+		query.Filter(priceFilterQuery)
+	}
+
+	es := connectors.BookConnectors.ElasticClient
+	ctx := context.Background()
+
+	searchResult, err := es.Client.Search().
+		Index(es.Index).
+		Type(es.DocType).
+		Query(query).
+		From(config.SearchResultFrom).
+		Size(config.SearchResultSize).
+		Pretty(true).
+		Do(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var books []models.Book
+
+	for _, hit := range searchResult.Hits.Hits {
+		var book models.Book
+		data, err := json.Marshal(hit.Source)
+		if err != nil {
+			return nil, err
+		}
+		json.Unmarshal(data, &book)
+		if err != nil {
+			return nil, err
+		}
+		books = append(books, book)
+	}
+
+	return books, nil
+
+}
+
+func (e ElasticBookController) Info() (map[string]interface{}, error) {
+
+	info := make(map[string]interface{})
+
+	es := connectors.BookConnectors.ElasticClient
+	ctx := context.Background()
+
+	sr, err := es.Client.Search().Aggregation("author_count", elastic.NewCardinalityAggregation().Field("author_name.keyword")).
+		Index(es.Index).
+		Type(es.DocType).
+		Size(0).
+		Do(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var authorsCount int
+	authors, found := sr.Aggregations.Terms("author_count")
+	if found {
+		for _, count := range authors.Aggregations {
+			if i, err := json.Marshal(&count); err != nil {
+				return nil, err
+			} else if authorsCount, err = strconv.Atoi(string(i)); err != nil {
+				return nil, err
+			}
+		}
+		info["author_count"] = authorsCount
+		info["books_count"] = sr.Hits.TotalHits
+
+	} else {
+		return nil, &elastic.Error{Status: http.StatusNotFound}
+	}
```

Comment:
I think that the cardinality aggregation return the number of different authors straight away ( https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-cardinality-aggregation.html)
So maybe you can just use the agg result as is:
```suggestion
	authorsCount, found := sr.Aggregations["author_count"]
	if found {
		info["author_count"] = authorsCount
		info["books_count"] = sr.Hits.TotalHits
	} else {
		return nil, &elastic.Error{Status: http.StatusNotFound}
	}
```
Just make sure I'm not completely off here :)

Reply:
its works! I have a question about this..